Addition to the Apple Globe key code to QMK

ChrisGVE opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

The change brings the Apple Globe Key as `KC_GLOBE` to QMK.

<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

The change is based mainly on @lordpixel23 [QMK patch](https://gist.github.com/lordpixel23/87498dc42e328eabdff6dd258a667efd) and is being discussed in depth in [QMK Apple Fn Key](https://gist.github.com/fauxpark/010dcf5d6377c3a71ac98ce37414c6c4?permalink_comment_id=4723100) created by @fauxpark.

Importantly, this change does not rely on VID/PID to be set on specific values.

In addition to the patch mentioned above, it introduces a new compilation flag, `GLOBE_KEY_ENABLE` such that:
 (a) the required `KEYBOARD_SHARE_EP` is enabled when the Globe keycode is used
 (b) keyboards or user layouts can make use of this definition for further customizations

The QMK documentation has been amended accordingly.

Certain key combinations are not yet working, though; in particular, 🌐︎+arrow or 🌐︎+FN do not work.

Thanks go to the input of many contributors, as this topic has been long discussed: @fauxpark, @HVR88, @drashna, and @lordpixel23.

<!--- Describe your changes in detail here. -->

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [x] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* #16651 issue was closed without solution

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

tzarc commented on data/constants/keycodes/keycodes_0.0.4_basic.hjson:
This needs to be in a 0.0.7 file, not 0.0.4, and there needs to be a plain `keycodes_0.0.7.hjson` with content `{}`.

After that's done, run `./util/regen.sh` and it'll sort the rest out.

tzarc commented on docs/keycodes.md:
Lot of unnecessary churn, here. Please limit the changes to what's actually needed, mirroring the existing formatting.

tzarc commented on docs/keycodes_basic.md:
Same here.

tzarc reviewed (changes requested):
I'll leave the rest of the review for the team -- they know more about the status of this stuff than what I do.

fauxpark commented on keyboards/zlant_xl/config.h.old:
Please remove this file.

fauxpark commented on keyboards/zlant_xl/rules.mk.old:
Please remove this file.

fauxpark commented on tmk_core/protocol.mk:
```diff
@@ -6,6 +6,13 @@ SRC +=	\
 MOUSE_SHARED_EP ?= yes
+
+ifeq ($(strip $(GLOBE_KEY_ENABLE)), yes)
+	OPT_DEFS += -DGLOBE_KEY_ENABLE
+	# The globe key require SHARED_EP
+	SHARED_EP_ENABLE = yes
```
```suggestion
    OPT_DEFS += -DGLOBE_KEY_ENABLE
    # The globe key requires SHARED_EP
    SHARED_EP_ENABLE = yes
```

fauxpark commented on tmk_core/protocol.mk:
```diff
@@ -6,6 +6,13 @@ SRC +=	\
 SHARED_EP_ENABLE = no
 MOUSE_SHARED_EP ?= yes
+
+ifeq ($(strip $(GLOBE_KEY_ENABLE)), yes)
+    OPT_DEFS += -DGLOBE_KEY_ENABLE
+    # The globe key requires SHARED_EP
+    SHARED_EP_ENABLE = yes
+endif
```
On closer inspection this doesn't really do anything since the keycode is implemented as just another Consumer usage under `EXTRAKEY_ENABLE`, which already sets `SHARED_EP_ENABLE = yes`.

Did you mean to set `KEYBOARD_SHARED_EP` instead? I can imagine the report descriptor containing the Globe key might be required to be on the same interface as the keyboard. I haven't played with this stuff in a while but that could get the F-keys working. Otherwise, this file doesn't need to be touched at all.

ChrisGVE replied:
Oh, you are right, my bad! Indeed, I mean to set KEYBOARD_SHARED_EP instead.
Let me correct that

fauxpark replied:
TBH I'm unsure if we truly need a `GLOBE_KEY_ENABLE` anyway as all it's doing is hiding what's actually going on - I think I would prefer to just call out `KEYBOARD_SHARED_EP` directly in the documentation, especially if the Globe key doesn't work at all without being on the same interface as the keyboard report.

ChrisGVE replied:
I agree this flag is pretty trivial, but I thought it would be more self-explanatory for users who only create layouts or even keyboard builders than the more obscure `KEYBOARD_ENABLE_EP`.

fauxpark replied:
The issue I see is that `GLOBE_KEY_ENABLE = yes` without `EXTRAKEY_ENABLE = yes` doesn't actually enable the Globe key by itself, but it would still combine the keyboard report descriptor into the shared endpoint. So the functionality here can potentially be non-obvious, and I'd rather be more explicit about it.

ChrisGVE replied:
I see. Well, I don't have a strong opinion here. I'm fine editing the documentation to make it more obvious for users. I would only need you to guide me where you'd like that done, and I can work on it tomorrow.

fauxpark commented on docs/keycodes.md:
```diff
@@ -211,13 +211,15 @@ See also: [Basic Keycodes](keycodes_basic)
 <sup>4. Holding Shift+Option allows for finer control of volume level.</sup><br/>
 <sup>5. Skips the entire track in iTunes when tapped, seeks within the current track when held.</sup><br/>
 <sup>6. WMP does not recognize the Rewind key, but both alter playback speed in VLC.</sup>
+<sup>7. The Globe key requires `GLOBE_KEY_ENABLE = yes` to be set in `rules.mk`. When pressed it will reveal the emoji picker, and it can be combined with other keys. Note however that Globe+F-keys or Globe+arrow do not work.</sup>
```
```suggestion
<sup>7. Requires `KEYBOARD_SHARED_EP = yes` in `rules.mk` or `usb.shared_endpoint.keyboard=true` in `keyboard.json`. Globe+F-keys/arrow keys do not work.</sup>
```
Specific functionality of the Globe key depends on the OS setting.
This could be made a little more concise by adding proper documentation+explanation for the shared endpoint stuff and linking to it instead... but that's a separate issue.

fauxpark commented on tmk_core/protocol.mk:
```diff
@@ -4,6 +4,12 @@ SRC +=	\
 	$(PROTOCOL_DIR)/usb_device_state.c \
 	$(PROTOCOL_DIR)/usb_util.c \
 
+ifeq ($(strip $(GLOBE_KEY_ENABLE)), yes)
+    OPT_DEFS += -DGLOBE_KEY_ENABLE
+    # The globe key requires KEYBOARD_SHARED_EP
+    KEYBOARD_SHARED_EP = yes
+endif
+
```
```suggestion
```

fauxpark commented:
Please undo the last merge commit.

fauxpark commented:
Now everything else is gone...

ChrisGVE commented:
Ok we should be back on track, sorry about the mishap.

avsej commented on tmk_core/protocol/report.h:
```diff
@@ -327,6 +327,8 @@ static inline uint16_t KEYCODE2CONSUMER(uint8_t key) {
         case KC_LAUNCHPAD:
             return AC_SOFT_KEY_LEFT;
+        case KC_GLOBE:
+            return AC_NEXT_KEYBOARD_LAYOUT_SELECT;
```
Looks like this KC_GLOBE does exactly that, and only suitable for switching layouts, not really sending Globe codes.

ChrisGVE replied:
Strictly speaking, you are right, but in practice, on a Mac, it operates as a key.
- pressed alone, it triggers the emoji picker (or what is configured in the Mac settings for the Keyboard)
- it can be used as a "modifier" with specific keys, for instance 
  - Globe+A to toggle the dock when hidden, 
  - Globe+C to toggle control center, 
  - Globe+F to toggle full screen, 
  - Globe+H to show the desktop, 
  - Globe+N to toggle Notification Center, 
  - Globe+Q launch notes.
Unfortunately, there are other cases where I have yet to have success. In particular, Globe+Arrow keys combinations, but I typically use 60% keyboards, so arrow keys are hard to test with those.

drashna replied:
If the keycode works on Windows/Linux too (but as layout select), then it would be worth having appropriate aliases too.

ChrisGVE replied:
I don't have a pure Windows or Linux machine, but I can investigate using a virtual machine. I'll see whether I have time tomorrow to try that.

fauxpark replied:
Windows does not seem to react to the `AC Next Keyboard Layout Select` usage, but I can see it coming in through Wireshark.

ChrisGVE replied:
Yes, I tried this morning while at work on Windows 10, and nada, no reaction to the key. I'll try Linux later; I suspect I'll see the same thing.

ChrisGVE replied:
Sorry for the delay. I have tried the globe key under a virtual Linux running on my Mac. But naively, I did not realize that the key would be handled by the Mac, and thus, I did not get any different behavior than for regular Mac apps. I would need someone with a Linux box to try natively. Sorry about that.

fauxpark replied:
If it helps, the kernel input handling code does know about this consumer usage, but I suppose it's up to the higher layers as to what happens with it:
https://elixir.bootlin.com/linux/v6.12.1/source/drivers/hid/hid-input.c#L1234

And you should be able to passthrough the USB device directly to the VM (at least on Virtualbox this is possible) to stop the host OS capturing it.

ChrisGVE replied:
You're right about the VM. Indeed, this is what happens when I connect to work via Citrix (the reason so many of my colleagues have keyboard problems). I'll try it again by configuring Parallels properly. I'll do a search, but by any chance, do you know any Linux software that would capture keyboard events like Karabiner on the Mac?

ChrisGVE replied:
I succeeded in passing all my keypresses to the Linux session running in VM. I've monitored the keypresses using `xev`, which is pretty basic but does the job. Like with Windows, the result is no report of any key being pressed when using either the Globe on its own or in combination with another key.

fauxpark replied:
I figured that might be the case, considering X keysyms only go up to 247 (for a total of 248 possible values); `KEY_KBD_LAYOUT_NEXT` has a value of 584. `evdev` might have better info, but even so, I'm not sure what the mechanism(s) would be, if any, to get the key to do something useful.

sigprof replied:
Actually [xkeyboard-config](https://gitlab.freedesktop.org/xkeyboard-config/xkeyboard-config) has [support for `KEY_KBD_LAYOUT_NEXT` in `symbols/inet`](https://gitlab.freedesktop.org/xkeyboard-config/xkeyboard-config/-/commit/2a0c538c518becc49aedd66dc21460418af1dcd8#f3d4a1d386e363c7fedc3c05285f0343c340e9cf_233_356) since `xkeyboard-config-2.33` (the `<I592>` keycode (`linux_input_keycode + 8`, as usual) gets mapped to `ISO_Next_Group`).  The problem is that the X protocol does not support keycode values above 255, therefore that mapping would work only in Wayland.

ChrisGVE replied:
Well, clearly, there is atm only a use case for macOS but it is a big one for Apple users. I'd wish to know how to make all the other combinations work but that's already a beginning that would benefit this group of users.

ChrisGVE commented:
Nothing on my end

drashna commented:
> Is there anything truly still open or can this land?

I can't speak for others, but I was waiting on @fauxpark, as he has more experience with the USB endpoint stuff, and the apple FN/globe key stuff. 

However, develop isn't merged until end of feb

fauxpark commented:
To be completely honest, I'm still hesitant to hit approve on this myself simply due to the fact the only known way to make this key fully functional is to spoof the VID/PID.

My initial goal way back when was to figure out how the Fn key worked as it pertained to the F-keys, and as I learned more about that it became clear that their behaviour changes based on the particular Apple keyboard model. So it seems to me that the Globe key in this current implementation will never really be able to do everything it can on a real device - by design.

By all means, if people are happy with this in its current state, go ahead. It's already got two green ticks. Great work and not your fault @ChrisGVE, but to me personally, it doesn't quite close the case, which is a shame.

lordpixel23 commented:
I use it every day as it is because it is already useful. It is frustrating it cannot be 100% but if we find a way to do that it can always be in some future PR.

tzarc commented:
Can we clearly document the full list of shortcomings before merge, then? A single footnote is a bit hard to link to when it the next person who comes along complains that it doesn’t work as per their Apple-branded device.

ChrisGVE commented:
> To be completely honest, I'm still hesitant to hit approve on this myself simply due to the fact the only known way to make this key fully functional is to spoof the VID/PID.
> 
> My initial goal way back when was to figure out how the Fn key worked as it pertained to the F-keys, and as I learned more about that it became clear that their behaviour changes based on the particular Apple keyboard model. So it seems to me that the Globe key in this current implementation will never really be able to do everything it can on a real device - by design.
> 
> By all means, if people are happy with this in its current state, go ahead. It's already got two green ticks. Great work and not your fault @ChrisGVE, but to me personally, it doesn't quite close the case, which is a shame.

This is really a shame because I am using this key on all my boards, and I have a few see my userspace repo without any problem. I don't have to change VIP/PID to make it work, whether I tap the key to obtain the emoji picker dialog or when I use it in combination with other keys (either as a single key or when using a hold-esc key + other key on HHKB).

I think I gave a list of cases where it does not work earlier in this PR, and the documentation could be changed to make it more obvious, but I still stand by its usefulness.

(My only challenge will be to port it back to the Keychron board Q60 Max, which is based on an earlier version of QMK...)

tzarc commented:
Have deferred this to next quarter due to git conflicts as well as internal discussion within the team as to if it should be merged.

In the meantime, as #24848 has been merged, I've added a quick-and-dirty equivalent as a community module [here](https://github.com/tzarc/qmk_modules/tree/main/globe_key).

ChrisGVE commented:
Thanks, @tzarc. I'll try the new feature soon, I've been in the weeds for quite some time...


In the meantime, we'll have to continue discussing adding this feature to the main branch for Q2. To start it, @fauxpark, @lordpixel23, @drashna, @avsej: From your point of view, what should be done to move forward on this topic? Can you spell out what an MVP would be acceptable/necessary for inclusion in the main branch?

ChrisGVE commented:
@avsej I agree with you: it’s a bummer. From what I could test, globe + alpha does work, but globe does not like modifiers and arrows, and I would love to be able to find out how to fix this. I unfortunately don’t have as much experience as you guys to go deeper, but I wouldn’t mind being guided if anyone has an idea.


> On 10 Mar 2025, at 5:08 PM, Sergey Avseyev ***@***.***> wrote:
> 
> 
> avsej
>  left a comment 
> (qmk/qmk_firmware#24661)
> @ChrisGVE my usecase is not really to switch layouts (I'm using just ⌘+space for that). For me it is more important to emulate globe (🌐︎) as it happens with built-in keyboard of the macbook. So that I can use it with default keybindings that it has for changing window layout since recent versions of MacOS, (for example, ⌃🌐︎↑ to fill current window in the top half of the screen).
> 
> I tested your patch, and it works well when it comes to main purpose of the globe 🌐︎-key when I just press-and-release it, but it does not work with 🌐︎-key combinations in other places of the system (ctrl+globe+up does not work).
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.
> 
>  <https://github.com/ChrisGVE> <https://github.com/qmk/qmk_firmware/pull/24661#issuecomment-2711105414> <https://github.com/notifications/unsubscribe-auth/AABULILGFFHV7R46KIIDE2T2TW2IVAVCNFSM6AAAAABSYP7CJ6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMJRGEYDKNBRGQ>
> 
> avsej
>  left a comment 
> (qmk/qmk_firmware#24661)
>  <https://github.com/qmk/qmk_firmware/pull/24661#issuecomment-2711105414>
> @ChrisGVE <https://github.com/ChrisGVE> my usecase is not really to switch layouts (I'm using just ⌘+space for that). For me it is more important to emulate globe (🌐︎) as it happens with built-in keyboard of the macbook. So that I can use it with default keybindings that it has for changing window layout since recent versions of MacOS, (for example, ⌃🌐︎↑ to fill current window in the top half of the screen).
> 
> I tested your patch, and it works well when it comes to main purpose of the globe 🌐︎-key when I just press-and-release it, but it does not work with 🌐︎-key combinations in other places of the system (ctrl+globe+up does not work).
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/qmk/qmk_firmware/pull/24661#issuecomment-2711105414>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AABULILGFFHV7R46KIIDE2T2TW2IVAVCNFSM6AAAAABSYP7CJ6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMJRGEYDKNBRGQ>.
> You are receiving this because you were mentioned.
>

tzarc commented:
I don’t have full context here, so forgive me if it’s already been discussed here or elsewhere — but given macOS maintains different capslock state for different keyboards, is macOS perhaps not considering the globe and the arrows as originating from the same board? Are the USB endpoints shared “correctly”?

avsej commented:
I gave up and reassigned some of the globe-specific shortcuts.
![image](https://github.com/user-attachments/assets/07bc5445-ef6c-43bf-9b28-bd20d3f6524f)
But it does not work in all applications (Neovide ignores it for some reason, although menu path look the same):
![image](https://github.com/user-attachments/assets/cc738a14-1c55-42f1-b6c4-2575d7dadc55)
![image](https://github.com/user-attachments/assets/f7bebc4f-c419-4391-ace5-e8921c651af3)

ChrisGVE commented:
@tzarc I'm only starting to explore this new feature. Will modules be included in the main repo, or will it be like qmk_userspace? IMHO that could be a good compromise for half-solutions such as the globe key in today's state, but I hope it becomes more available to other users. I read the new module doc, and I was confused.

tzarc commented:
> @tzarc I'm only starting to explore this new feature. Will modules be included in the main repo, or will it be like qmk_userspace? IMHO that could be a good compromise for half-solutions such as the globe key in today's state, but I hope it becomes more available to other users. I read the new module doc, and I was confused.

Modules may be submitted to QMK, but in the case above I was more referring to their use as a third-party importable module _today_, while we work out exactly what we should do with it with respect to QMK itself.

lordpixel23 commented:
I note ⌃🌐︎R, ⌃🌐︎F and ⌃🌐︎C do work, which means the main gap is the arrow keys in particular. Here's the Karabiner log for a successful use of ⌃🌐︎↑ etc on my internal keyboard

```json
[
 {
    "type": "down",
    "name": {"apple_vendor_top_case_key_code":"keyboard_fn"},
    "usagePage": "255 (0x00ff)",
    "usage": "3 (0x0003)",
    "misc": "flags fn"
  },
  {
    "type": "down",
    "name": {"key_code":"left_control"},
    "usagePage": "7 (0x0007)",
    "usage": "224 (0x00e0)",
    "misc": "flags fn,left_control"
  },
  {
    "type": "down",
    "name": {"key_code":"down_arrow"},
    "usagePage": "7 (0x0007)",
    "usage": "81 (0x0051)",
    "misc": "flags fn,left_control"
  },
  {
    "type": "up",
    "name": {"key_code":"down_arrow"},
    "usagePage": "7 (0x0007)",
    "usage": "81 (0x0051)",
    "misc": "flags fn,left_control"
  },
  {
    "type": "down",
    "name": {"key_code":"up_arrow"},
    "usagePage": "7 (0x0007)",
    "usage": "82 (0x0052)",
    "misc": "flags fn,left_control"
  },
  {
    "type": "up",
    "name": {"key_code":"up_arrow"},
    "usagePage": "7 (0x0007)",
    "usage": "82 (0x0052)",
    "misc": "flags fn,left_control"
  },
  {
    "type": "up",
    "name": {"key_code":"left_control"},
    "usagePage": "7 (0x0007)",
    "usage": "224 (0x00e0)",
    "misc": "flags fn"
  },
  {
    "type": "up",
    "name": {"apple_vendor_top_case_key_code":"keyboard_fn"},
    "usagePage": "255 (0x00ff)",
    "usage": "3 (0x0003)",
    "misc": ""
  },
 ]
```
And here's the same thing on the external keyboard with QMK

```json
[

  {
    "type": "down",
    "name": {"key_code":"left_control"},
    "usagePage": "7 (0x0007)",
    "usage": "224 (0x00e0)",
    "misc": "flags left_control"
  },
  {
    "type": "down",
    "name": {"key_code":"up_arrow"},
    "usagePage": "7 (0x0007)",
    "usage": "82 (0x0052)",
    "misc": "flags left_control"
  },
  {
    "type": "up",
    "name": {"key_code":"up_arrow"},
    "usagePage": "7 (0x0007)",
    "usage": "82 (0x0052)",
    "misc": "flags left_control"
  },
  {
    "type": "down",
    "name": {"key_code":"down_arrow"},
    "usagePage": "7 (0x0007)",
    "usage": "81 (0x0051)",
    "misc": "flags left_control"
  },
  {
    "type": "up",
    "name": {"key_code":"down_arrow"},
    "usagePage": "7 (0x0007)",
    "usage": "81 (0x0051)",
    "misc": "flags left_control"
  },
  {
    "type": "up",
    "name": {"key_code":"left_control"},
    "usagePage": "7 (0x0007)",
    "usage": "224 (0x00e0)",
    "misc": ""
  },
]
```
As you can see the fn or globe key simply doesn't register here. Though, in Karabiner Events is does in fact show up in the Unknown Events tab
```json
[
  {
    "value": "1",
    "usagePage": "12 (0x000c)",
    "usage": "669 (0x029d)"
  },
  {
    "value": "0",
    "usagePage": "12 (0x000c)",
    "usage": "669 (0x029d)"
  },
]
```

It is interesting to note that when I do globe-ctrl-C on the QMK keyboard, it looks like it _should_ be broken because the fn is missing (compared to the built in keyboard) but, in fact, this actually *DOES* work.
```json
[
  {
    "type": "down",
    "name": {"key_code":"left_control"},
    "usagePage": "7 (0x0007)",
    "usage": "224 (0x00e0)",
    "misc": "flags left_control"
  },
  {
    "type": "down",
    "name": {"key_code":"c"},
    "usagePage": "7 (0x0007)",
    "usage": "6 (0x0006)",
    "misc": "flags left_control"
  },
  {
    "type": "up",
    "name": {"key_code":"c"},
    "usagePage": "7 (0x0007)",
    "usage": "6 (0x0006)",
    "misc": "flags left_control"
  },
  {
    "type": "up",
    "name": {"key_code":"left_control"},
    "usagePage": "7 (0x0007)",
    "usage": "224 (0x00e0)",
    "misc": ""
  },
]
```
From this we can see that Karabiner Events at least sees these as different keys. The operating system appears to treat them as equivalent for some purposes but not others.

ChrisGVE commented:
Ok let me take care of that today.Sent from my iPhoneOn 19 Mar 2025, at 01:53, Andrew Thompson ***@***.***> wrote:﻿
  Reading back through this thread the only concrete request I see is from tzarc who said:

Can we clearly document the full list of shortcomings before merge, then? A single footnote is a bit hard to link to when it the next person who comes along complains that it doesn’t work as per their Apple-branded device.
—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>



lordpixel23 left a comment (qmk/qmk_firmware#24661)
Reading back through this thread the only concrete request I see is from tzarc who said:

Can we clearly document the full list of shortcomings before merge, then? A single footnote is a bit hard to link to when it the next person who comes along complains that it doesn’t work as per their Apple-branded device.


—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

bintoro commented:
I wouldn’t be surprised if the window management shortcuts involving arrow keys not working were in fact a macOS bug.

The Fn key (as a feature of Apple keyboards specifically) has traditionally had the following uses:
- control the behaviour of the function key row
- make the arrow keys function as Home/End/PgUp/PgDn
- make the Return key function as Enter (mostly irrelevant today).

The Globe key is a more recent invention. It’s available to third-party vendors to implement, but on Apple keyboards it merges with the Fn key.

Various system-level shortcuts involving the Globe key have sprouted up in recent years. In contrast to the Fn key combos, which exist to enhance the functionality of Apple keyboards, the Globe shortcuts are bona fide shortcuts that use Globe as a modifier because its use is disallowed in app-level shortcuts.

It makes no sense that a random subset of the Globe category (⌃🌐︎ + arrows) should require Fn instead of Globe. Together with the fact that these shortcuts were only introduced in macOS Sequoia, this leads me to believe that they may eventually begin to work on third-party keyboards as well.

Given that the Fn-specific functions and the Globe shortcuts are two different things, I don’t think the Globe key should be omitted from QMK just because it doesn’t do what it’s not even supposed to do.

I didn’t see anything surprising in the Karabiner log. Karabiner doesn’t currently recognize the Globe code. I’ve been meaning to open an issue to have it added, but if one is using QMK and Karabiner, there isn’t really a need for Globe, as you can have the fully functional Fn instead.

(In the absence of Globe support, I have assigned F20 to the physical Fn key on my QMK keyboard. F20 is not usable in macOS but can be used as input to Karabiner, so it’s perfect for this.)

ChrisGVE commented:
I have added more documentation and warnings about the partial support for the Globe key. Could you double-check that I have modified the correct documentation files and that the links work? Sorry for the trouble, but this is the first time I have made such a significant change to the QMK doc.

ChrisGVE commented:
Hello all, the PR has been marked as stale, but I had made the changes that were requested at the time. We have all been busy with other things. Still, at this point, I see that my local develop version has diverged quite a bit from the current branch (with a lot of files that I have not modified), so before I start spending time refreshing my branch, I'd like to know whether the team is prepared to accept this feature into the main branch? I hope it does but it is up to you 😉

volen23 commented:
Why don't you publish this as a module yourself?

tzarc commented:
> Why don't you publish this as a module yourself?

I've already done so; you can get it here: https://github.com/tzarc/qmk_modules/tree/main/globe_key

I think I'm going to just mark this PR as "on-hold" for the moment -- having a module available should suffice until someone works out how to get true Fn/Globe key compatibility on macOS.

arakashic commented:
Just to share an alternative solution. I map QK_PROGRAMMABLE_BUTTON_19 to the Globe key in software (via Karabiner Element). It gives the full functionality of the Globe/FN key without having to change the QMK firmware at all.

Here is the config:
```json
{
    "description": "Maps QK_PROGRAMMABLE_BUTTON_19 to Globe/FN",
    "manipulators": [
        {
            "from": { "pointing_button": "button19" },
            "to": [{ "key_code": "fn" }],
            "type": "basic"
        }
    ]
}
```
